File: internal/cli/console.go
// Package cli — VM console access commands — connect, state, kill
package cli

import (
	"context"
	"fmt"
	"os"

	"mvmctl/internal/cli/common"
	"mvmctl/pkg/api"

	"github.com/spf13/cobra"
)

func NewConsoleCmd(consoleAPI api.ConsoleAPI) *cobra.Command {
	var state bool
	var kill bool

	cmd := &cobra.Command{
		Use:   "console [vm-name]",
		Short: "VM console access",
		Long: `Attach to a VM console.

Provide a VM identifier (name, ID prefix, IP, or MAC address) as the
positional argument.

Press Ctrl+X then D to detach from the console.

Use --state to show the console relay state without attaching.
Use --kill to stop the console relay.`,
		// Python uses no_args_is_help=True on the Typer group, so running
		// "mvm console" with no args prints help text instead of an error.
		Args:              cobra.MaximumNArgs(1),
		ValidArgsFunction: completeVMNames,
		RunE: func(cmd *cobra.Command, args []string) error {
			// Simulate no_args_is_help=True: show help when no arguments given.
			if len(args) == 0 {
				return cmd.Help()
			}
			identifier := args[0]

			if state {
				return showConsoleState(consoleAPI, cmd.Context(), identifier)
			}
			if kill {
				return killConsoleRelay(consoleAPI, cmd.Context(), identifier)
			}
			return attachToConsole(consoleAPI, cmd, identifier)
		},
	}

	cmd.Flags().BoolVar(&state, "state", false, "Show console state without attaching")
	cmd.Flags().BoolVar(&kill, "kill", false, "Kill the console relay")

	return cmd
}

func showConsoleState(consoleAPI api.ConsoleAPI, ctx context.Context, identifier string) error {
	state, err := consoleAPI.ConsoleGetState(ctx, identifier)
	if err != nil {
		// Python's _show_console_state does NOT catch exceptions from
		// ConsoleOperation.get_state() — they propagate to @handle_errors,
		// which calls mvm_cli.error(str(e)) then raises typer.Exit(1).
		// In Go, print the error and return it (SilenceErrors on the root
		// command prevents Cobra from double-printing).
		common.Cli.Error(err.Error())
		return err
	}

	status := "stopped"
	if state.Running {
		status = "running"
	}
	common.Cli.Info(fmt.Sprintf("Console for '%s': %s", identifier, status))

	if state.PID != nil && *state.PID != 0 {
		common.Cli.Info(fmt.Sprintf("  PID: %d", *state.PID))
	}

	if state.SocketPath != "" {
		common.Cli.Info(fmt.Sprintf("  Socket: %s", state.SocketPath))
	}
	common.Cli.Info(fmt.Sprintf("Console for '%s': %s", identifier, status))
	return nil
}

func killConsoleRelay(consoleAPI api.ConsoleAPI, ctx context.Context, identifier string) error {
	err := consoleAPI.ConsoleKill(ctx, identifier)
	if err != nil {
		common.Cli.Error(err.Error())
		return err
	}

	common.Cli.Success(fmt.Sprintf("Stopped: %s", identifier))
	return nil
}

func attachToConsole(consoleAPI api.ConsoleAPI, cmd *cobra.Command, identifier string) error {
	info, err := consoleAPI.ConsoleGetConnectionInfo(cmd.Context(), identifier)
	if err != nil {
		// Python: get_connection_info raises MVMError which propagates to
		// @handle_errors, which calls mvm_cli.error(str(e)) then typer.Exit(1).
		// In Go, print the error and return it.
		common.Cli.Error(err.Error())
		return err
	}

	common.Cli.Info(fmt.Sprintf("Attaching to console of '%s'...", info.VMName))
	common.Cli.Info("Press Ctrl+X then D to detach")

	err = consoleAPI.ConsoleAttachConsole(cmd.Context(), info.SocketPath, os.Stdin, os.Stdout)
	if err == nil {
		// Python's _attach_to_console: mvm_cli.info("\nDetached from console")
		common.Cli.Info("\nDetached from console")
	} else {
		// Python's socket connection failure is handled inside _connect_socket
		// (mvm_cli.error(f"Console relay connection failed: {e}")) and then
		// _attach_to_console checks for None and prints "Console relay connection
		// failed" before raising typer.Exit(1). MVMErrors during _interact are
		// caught and printed via mvm_cli.error(str(e)). In Go, InteractiveAttach
		// returns these errors so we print and return them here.
		common.Cli.Error(err.Error())
	}
	return err
}

File: legacy/python/src/mvmctl/cli/console.py
"""VM console access commands - connect, state, kill."""

from __future__ import annotations

import select
import socket
import sys
import termios
import tty

import typer

from mvmctl.cli._completion import _complete_vm_names
from mvmctl.constants import CONST_CONSOLE_SOCKET_TIMEOUT_S
from mvmctl.exceptions import MVMError
from mvmctl.utils.cli import handle_errors, mvm_cli

console_app = typer.Typer(
    help="VM console access",
    no_args_is_help=True,
    add_completion=False,
    context_settings={"allow_interspersed_args": True},
)


@console_app.callback(invoke_without_command=True)
@handle_errors
def console(
    ctx: typer.Context,
    identifier: str = typer.Argument(
        ...,
        help="VM name, ID, IP, or MAC address",
        autocompletion=_complete_vm_names,
    ),
    state: bool = typer.Option(
        False, "--state", help="Show console state without attaching"
    ),
    kill: bool = typer.Option(False, "--kill", help="Kill the console relay"),
) -> None:
    """
    Attach to a VM console.

    Provide a VM identifier (name, ID prefix, IP, or MAC address) as the
    positional argument.

    Press Ctrl+X then D to detach from the console.
    """
    if ctx.invoked_subcommand is not None:
        return

    if state:
        _show_console_state(identifier)
    elif kill:
        _kill_console_relay(identifier)
    else:
        _attach_to_console(identifier)


def _show_console_state(identifier: str) -> None:
    """Display console relay state for a VM."""
    from mvmctl.api import ConsoleOperation

    state_dict = ConsoleOperation.get_state(identifier)

    status = "running" if state_dict["running"] else "stopped"
    mvm_cli.info(f"Console for '{identifier}': {status}")
    if state_dict["pid"]:
        mvm_cli.info(f"  PID: {state_dict['pid']}")
    if state_dict["socket_path"]:
        mvm_cli.info(f"  Socket: {state_dict['socket_path']}")


def _kill_console_relay(identifier: str) -> None:
    """Kill the console relay for a VM."""
    from mvmctl.api import ConsoleOperation

    result = ConsoleOperation.kill(identifier)

    if result.status == "success":
        mvm_cli.success(f"Stopped: {identifier}")
    elif result.status == "skipped":
        mvm_cli.error(f"Console relay not running: {identifier}")
        raise typer.Exit(1)
    else:
        mvm_cli.error(result.message or f"Stop failed: {identifier}")
        raise typer.Exit(1)


def _attach_to_console(identifier: str) -> None:
    """Attach to VM console interactively."""
    from mvmctl.api import ConsoleOperation

    attach_info = ConsoleOperation.get_connection_info(identifier)

    mvm_cli.info(f"Attaching to console of '{attach_info.vm_name}'...")
    mvm_cli.info("Press Ctrl+X then D to detach")

    sock = _connect_socket(attach_info.socket_path)
    if sock is None:
        mvm_cli.error("Console relay connection failed")
        raise typer.Exit(1)

    old_tty = None
    try:
        old_tty = termios.tcgetattr(sys.stdin)
        tty.setraw(sys.stdin.fileno())
        _interact(sock)
        mvm_cli.info("\nDetached from console")
    except KeyboardInterrupt:
        pass
    except MVMError as e:
        mvm_cli.error(str(e))
    finally:
        if old_tty is not None:
            termios.tcsetattr(sys.stdin, termios.TCSADRAIN, old_tty)
        try:
            sock.close()
        except OSError:
            pass


def _connect_socket(socket_path: str) -> socket.socket | None:
    """Connect to the console relay Unix socket."""
    try:
        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sock.settimeout(CONST_CONSOLE_SOCKET_TIMEOUT_S)
        sock.connect(socket_path)
        sock.setblocking(False)
        return sock
    except (OSError, ConnectionRefusedError, FileNotFoundError) as e:
        mvm_cli.error(f"Console relay connection failed: {e}")
        return None


def _interact(sock: socket.socket) -> None:
    """Run interactive console I/O loop."""
    buffer_size = 4096
    input_buffer = bytearray()

    while True:
        ready, _, _ = select.select([sys.stdin, sock], [], [], 0.05)

        if sock in ready:
            try:
                data = sock.recv(buffer_size)
                if data:
                    sys.stdout.buffer.write(data)
                    sys.stdout.flush()
                else:
                    return
            except (BlockingIOError, InterruptedError):
                pass
            except (OSError, ConnectionResetError):
                return

        if sys.stdin in ready:
            char = sys.stdin.buffer.read(1)
            if not char:
                return

            input_buffer.extend(char)

            if len(input_buffer) >= 2:
                if bytes(input_buffer[-2:]) == b"\x18d":
                    if len(input_buffer) > 2:
                        leftover = bytes(input_buffer[:-2])
                        _try_send(sock, leftover)
                    return

            if input_buffer[0:1] != b"\x18":
                to_send = bytes(input_buffer)
                if to_send:
                    _try_send(sock, to_send)
                input_buffer = bytearray()
            elif len(input_buffer) >= 2:
                if bytes(input_buffer) != b"\x18d":
                    to_send = bytes(input_buffer)
                    if to_send:
                        _try_send(sock, to_send)
                    input_buffer = bytearray()


def _try_send(sock: socket.socket, data: bytes) -> None:
    try:
        sock.sendall(data)
    except (OSError, BrokenPipeError, ConnectionResetError):
        pass
